include/swift/Sema/Requests.h

```diff
@@ -0,0 +1,121 @@
+//===--- Requests.h - Type Checking Requests -------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines type checking requests.
+//
+//===----------------------------------------------------------------------===//
+#ifndef SWIFT_SEMA_REQUESTS_H
+#define SWIFT_SEMA_REQUESTS_H
+
+#include "swift/AST/Type.h"
+#include "swift/AST/Evaluator.h"
+#include "llvm/ADT/Hashing.h"
+
+namespace swift {
+
+struct TypeLoc;
+
+/// Request the type from the ith entry in the inheritance clause for the
+/// given declaration.
+struct InheritedTypeRequest {
+  // Data.
+  llvm::PointerUnion<TypeDecl *, ExtensionDecl *> decl;
+  unsigned index;
+
+  /// Retrieve the TypeLoc for this inherited type.
+  TypeLoc &getTypeLoc() const;
+
+public:
+  using OutputType = Type;
+
+  // Evaluation.
+  Type operator()(Evaluator &evaluator) const;
+
+  // Cycle handling
+  Type breakCycle() const { return Type(); }
+  void diagnoseCycle(DiagnosticEngine &diags) const { }
+  void noteCycleStep(DiagnosticEngine &diags) const { }
+
+  // Caching
+  static const bool isEverCached = true;
+  static const bool hasExternalCache = true;
+
+  bool isCached() const { return true; }
+  bool isInFlight() const;
+  void setInFlight() const;
+  Optional<Type> getCachedResult() const;
+  void cacheResult(Type value) const;
+
+  // Value type.
+  friend bool operator==(const InheritedTypeRequest &lhs,
+                         const InheritedTypeRequest &rhs) {
+    return lhs.decl == rhs.decl && lhs.index == rhs.index;
+  }
+
+  friend bool operator!=(const InheritedTypeRequest &lhs,
+                         const InheritedTypeRequest &rhs) {
+    return !(lhs == rhs);
+  }
+
+  friend hash_code hash_value(const InheritedTypeRequest &req) {
+    return llvm::hash_combine(req.decl.getOpaqueValue(), req.index);
+  }
+};
+
+SWIFT_TYPEID(TypeChecker, InheritedTypeRequest, 1);
```

Comment:
Doing these individually by hand seems prone to mistakes and problems with merge conflicts? Is it worth having (yet another) .def file?

Reply:
Yeah, I will do something here. It's an annoying pile of preprocessor metaprogramming.